ReactiveObjC/RACSignal+Operations.m

```diff
@@ -846,22 +846,57 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
 
 - (RACSignal *)switchToLatest {
   return [[RACSignal createSignal:^(id<RACSubscriber> subscriber) {
-    RACMulticastConnection *connection = [self publish];
+    __block BOOL outerCompleted = NO;
+    __block BOOL innerCompleted = NO;
+    __block NSUInteger innerSignalIndex = 0;
 
-    RACDisposable *subscriptionDisposable = [[connection.signal
-      flattenMap:^(RACSignal *x) {
-        NSCAssert(x == nil || [x isKindOfClass:RACSignal.class], @"-switchToLatest requires that the source signal (%@) send signals. Instead we got: %@", self, x);
+    RACSerialDisposable *innerDisposable = [[RACSerialDisposable alloc] init];
+    NSLock *lock = [[NSLock alloc] init];
+
+    RACDisposable *outerDisposable = [self subscribeNext:^(RACSignal *signal) {
+      NSCAssert(signal == nil || [signal isKindOfClass:RACSignal.class], @"-switchToLatest expects signals as values, instead we got: %@", signal);
+      signal = signal ?: [RACSignal empty];
 
-        // -concat:[RACSignal never] prevents completion of the receiver from
-        // prematurely terminating the inner signal.
-        return [x takeUntil:[connection.signal concat:[RACSignal never]]];
-      }]
-      subscribe:subscriber];
+      [innerDisposable.disposable dispose];
+
+      [lock lock];
+      innerCompleted = NO;
+      ++innerSignalIndex;
+      NSUInteger signalIndex = innerSignalIndex;
+      [lock unlock];
+
+      RACDisposable *disposable = [signal subscribeNext:^(id x) {
+        [subscriber sendNext:x];
+      } error:^(NSError *error) {
+        [subscriber sendError:error];
+      } completed:^{
+        [lock lock];
+        if (innerSignalIndex == signalIndex) {
+          if (outerCompleted) {
+            [subscriber sendCompleted];
+          } else {
+            innerCompleted = YES;
+          }
+        }
+        [lock unlock];
+      }];
+
+      [innerDisposable swapInDisposable:disposable];
```

Comment:
minor: I'm not sure that there's a concrete benefit to use a serial disposable over just a __block variable, since you're already disposing of the internal disposable beforehand - you're just incurring extra cost by using the atomic protections.

Reply:
How would you refer the `innerDisposable` otherwise?

Reply:
That wouldn't change.

See [this commit](https://github.com/Lightricks/ReactiveObjC/commit/6945e8210d4044a1e4d246e227c9acc2ad086544) for what I mean.